dll/win32/shell32/CDefView.cpp

```diff
@@ -3069,11 +3120,190 @@ HRESULT WINAPI CDefView::AddPropertySheetPages(DWORD dwReserved, LPFNADDPROPSHEE
+{
+    HRESULT hr = Read(pS, &cols, FIELD_OFFSET(PERSISTCOLUMNS, Columns));
+    if (FAILED(hr))
+        return hr;
+    if (cols.Signature != PERSISTCOLUMNS::SIG || cols.Count > cols.MAXCOUNT)
+        return HResultFromWin32(ERROR_INVALID_DATA);
+    return Read(pS, &cols.Columns, sizeof(*cols.Columns) * cols.Count);
+}
+
+HRESULT CDefView::LoadViewState()
+{
+    PERSISTCLASSICVIEWSTATE cvs;
+    PERSISTCOLUMNS cols;
+    CComPtr<IStream> pS;
+    CComPtr<IPropertyBag> pPB;
+    bool fallback = false;
+    HRESULT hrColumns = E_FAIL;
+    HRESULT hr = IUnknown_QueryServicePropertyBag(m_pShellBrowser, SHGVSPB_FOLDER, IID_PPV_ARG(IPropertyBag, &pPB));
+    if (SUCCEEDED(hr))
+    {
+        DWORD data;
+        if (FAILED(hr = SHPropertyBag_ReadDWORD(pPB, L"Mode", &data)))
+            goto loadfallback;
+        cvs.FolderSettings.ViewMode = data;
+        cvs.FolderSettings.fFlags = ReadDWORD(pPB, L"FFlags", FWF_NOGROUPING);
+        data = ReadDWORD(pPB, L"Sort", ~0ul);
+        cvs.SortColId = data != ~0ul ? (WORD)data : LISTVIEW_SORT_INFO::UNSPECIFIEDCOLUMN;
+        cvs.SortDir = (INT8)ReadDWORD(pPB, L"SortDir", 1);
+        if (SUCCEEDED(hrColumns = SHPropertyBag_ReadStream(pPB, L"ColInfo", &pS)))
+            hrColumns = LoadColumnsStream(cols, pS);
+    }
+    else
+    {
+        if (FAILED(hr = (m_pShellBrowser ? m_pShellBrowser->GetViewStateStream(STGM_READ, &pS) : E_UNEXPECTED)))
+        {
+        loadfallback:
+            hr = GetDefaultViewStream(STGM_READ, &pS);
+            fallback = true;
+        }
+        if (FAILED(hr) || FAILED(hr = Read(pS, &cvs, sizeof(cvs))))
+            return hr;
+        if (cvs.Signature != cvs.SIG)
+            return HResultFromWin32(ERROR_INVALID_DATA);
+        hrColumns = LoadColumnsStream(cols, pS);
+    }
+    m_FolderSettings.ViewMode = cvs.FolderSettings.ViewMode;
+    m_FolderSettings.fFlags &= ~cvs.VALIDFWF;
+    m_FolderSettings.fFlags |= cvs.FolderSettings.fFlags & cvs.VALIDFWF;
+    if (SUCCEEDED(hrColumns))
+    {
+        BOOL failed = FALSE;
+        if ((m_LoadColumnsList = DPA_Create(cols.Count)) != NULL)
+        {
+            for (UINT i = 0; i < cols.Count; ++i)
+            {
+                failed |= !DPA_SetPtr(m_LoadColumnsList, i, (void*)(SIZE_T)cols.Columns[i]);
+            }
+        }
+        if (failed || !cols.Count)
+        {
+            DPA_Destroy(m_LoadColumnsList);
+            m_LoadColumnsList = NULL;
+        }
+    }
+    m_sortInfo.bLoadedFromViewState = !fallback && m_LoadColumnsList && cvs.SortColId != LISTVIEW_SORT_INFO::UNSPECIFIEDCOLUMN;
+    m_sortInfo.bColumnIsFolderColumn = TRUE;
+    m_sortInfo.Direction = cvs.SortDir > 0 ? 1 : -1;
+    m_sortInfo.ListColumn = cvs.SortColId;
+    return hr;
+}
+
+HRESULT CDefView::SaveViewState(IStream *pStream)
+{
+    int sortcol = MapListColumnToFolderColumn(m_sortInfo.ListColumn);
+    PERSISTCLASSICVIEWSTATE cvs;
+    cvs.SortColId = sortcol >= 0 ? (WORD)sortcol : 0;
```

Comment:
Why not simply like that? I thought that on several places. This looks much simpler to me:
```suggestion
    cvs.SortColId = (WORD)sortcol;
```
Do I miss anything?

Reply:
Yes, it is negative on error.

Reply:
You could do `max((WORD)sortcol, 0)`.

Reply:
> You could do `max((WORD)sortcol, 0)`.

The WORD cast in this case would mean 0xffff for -1 and that is larger than 0?

Besides, I don't think it makes the code clearer. I prefer: `valid ? value : default`.

Reply:
`max` is defined as 
`#define MAX(a,b) ((a) > (b) ? (a) : (b))`
So no, it wouldn't be evaluated as 'larger'.
I do think using `max` is cleaner, but it isn't a big issue.

Reply:
Fair enough. I forgot for a second that WORD is typedef'd as an unsigned.